dhall/src/Dhall/Repl.hs

```diff
@@ -579,9 +580,9 @@ greeter =
   in liftIO (putStrLn message)
 
 
-dontCrash :: ( MonadIO m, Haskeline.MonadException m ) => m () -> m ()
+dontCrash :: ( MonadCatch m, MonadIO m ) => m () -> m ()
 dontCrash m =
-  Haskeline.catch
+  Control.Monad.Catch.catch
```

Comment:
I'm not very keen on this, but it might be possible to preserve compatibility with older versions of haskeline and repline with a bit of CPP.

If the hnix release takes a long time, that might be worth looking into…

Reply:
@sjakobi: You will need to support older versions of `haskeline` in order to support older versions of GHC.  `haskeline` is one of GHC's "boot libraries" (https://gitlab.haskell.org/ghc/ghc/-/wikis/commentary/libraries/version-history)

Reply:
`haskeline` is upgradeable – I can compile it fine with e.g. 8.6.5. Do you have different compatibility issues in mind?

Reply:
Never mind.  For some reason I thought `stack` could not upgrade boot libraries, but apparently `haskeline` is an exception

Reply:
@sjakobi: It looks like you might need to support older versions of Haskeline due to Nix.  It appears to misbehave in the following way when upgrading `haskeline`:

```
Warning:
    This package indirectly depends on multiple versions of the same package. This is very likely to cause a compile failure.
      package hnix (hnix-0.7.1) requires haskeline-0.7.4.2
      package repline (repline-0.3.0.0-Apjg8sKpeyJ3V54YeFv2Wr) requires haskeline-0.8.0.0-2fxBQp1muOsBHoaEfLruXE
```

Reply:
@Gabriel439 Shouldn't that error go away when we update `hnix` once https://github.com/haskell-nix/hnix/pull/540 is merged?

Reply:
@sjakobi: I'm not sure, but my guess is: probably not.  My rough understanding of what is going on is:

* The GHC built by Nix comes with boot libraries (e.g. `haskeline-0.7.4.2`)
* You can still build a newer `haskeline` (e.g. `haskeline-0.8.0.0`) and even build packages that depend on that newer `haskeline` (e.g. `repline-0.3.0.0`), but somehow things go awry when the newer `haskeline` becomes a transitive dependency because then the newer `haskeline` conflicts with the older `haskeline` included with GHC

So I think the problem would only disappear if we can figure out to prevent Nix from using the older `haskeline` at all.

Reply:
Ok, that sounds tricky.

I guess the easiest way to backward compatibility would be to avoid the `MonadException` / `MonadCatch` constraints, and to use our concrete `Repl` type instead:

```
type Repl = Repline.HaskelineT (State.StateT Env IO)
```

Do you think that could work?

Reply:
@sjakobi: Yeah, that seems fine